### ISSUE TITLE
jigsawpy nearing release

### ISSUE THREAD

**Issue by xylar** [MPAS-Dev/MPAS-Model#296]
[jigsawpy](https://github.com/dengwirda/jigsaw-geo-python/tree/dev) is close to being ready to release.  It would be good for us to convert at least a few of our test cases to use the new python package instead of the current Matlab interface so we find out if there are any issues.

Anaconda (or at least the conda-forge channel) isn't updating the list of available packages right now.  As soon as that issue is resolved and the new version of `jigsaw` (v0.9.11) is available, I will make a test COMPASS environment on LANL IC that includes `jigsawpy`.  We should, then, be able to use this environment to try moving some (and eventually all!) COMPASS test cases away from jigsaw-geo-matlab.

**Comment by xylar** [comment]
I have created a test environment on LANL IC with `jigsawpy`.  It can be activated with:
```
source /usr/projects/climate/SHARED_CLIMATE/anaconda_envs/base/etc/profile.d/conda.sh
conda activate compass_jigsawpy_py3.7
```

I am off to the AntClim21 meeting tomorrow so I won't have time to work on the `jigsawpy` conversion this week.  @mark-petersen, @sbrus89 or @pwolfram, if any of you have time to work on this, it could be very helpful.  If we discover bugs or issues right now, they can be included in the first release of `jigsawpy`.  If we wait to test, it will likely slow the process down since we'll be relying on @dengwirda's timetable for doing a future release.  Likely not a big deal but I think it's still best to get involved early.

**Comment by dengwirda** [comment]
Yes, very happy for feedback here -- thanks to @xylar for enduring my `Python` learning curve!

`jigsawpy` aims to replace the `OCTAVE` step in the current workflow. The `Python` and `MATLAB` interfaces aim to be very similar, so in many cases the same function calls, variable names, etc are available.

I'll also reach out to other user groups (`CSIRO`, `AWI`, etc) for additional feedback.

**Comment by xylar** [comment]
@mark-petersen, @sbrus89 and @pwolfram, I've worked on this when I should be doing other things.  Please use this branch as a starting point for the conversion to `jigsawpy`.  Note: I haven't debugged this *at all* so it's just a pretty complete sketch of what needs to happen.
https://github.com/xylar/MPAS-Model/tree/ocean/update_to_jigsawpy

**Comment by sbrus89** [comment]
@xylar, Awesome! Thanks for getting the ball rolling on this. I'll try to take a look at it this week.

**Comment by pwolfram** [comment]
Thanks @sbrus89, this would be really helpful!

**Comment by sbrus89** [comment]
@xylar, I worked on this a little bit today and made some progress. I have a couple minor commits that I'd like to push to your fork. Would you mind giving me access to do that?

**Comment by sbrus89** [comment]
Jigsaw runs through after making these changes I mentioned above, but right now it doesn't seem to save this file `opts.mesh_file = 'mesh-MESH.msh'` which we specify in `jigsaw_driver.py` here:

https://github.com/xylar/MPAS-Model/blob/26ef9cfb0157efeacc250386747b3db1034f2508/testing_and_setup/compass/ocean/jigsaw_to_MPAS/jigsaw_driver.py#L24-L29

As a result, the `triangle_jigsaw_to_netcdf.py` step fails because it expects a `mesh-MESH.msh` file.

**Comment by sbrus89** [comment]
As far as I can tell these are the only two files that Jigsaw is outputting are:

`opts.hfun_file = 'mesh-HFUN.msh'`
`opts.geom_file = 'mesh.msh'`

@dengwirda, do you know why this might be the case? Maybe we aren't specifying them correctly in `jigsaw_driver.py`

**Comment by dengwirda** [comment]
Thanks for looking at this @sbrus89. 

TL; DR: I think changing `jigsawpy.lib.jigsaw( ... )` to `jigsawpy.cmd.jigsaw(opts)` in `jigsaw_driver.py` should hopefully fix this.

~

It seems that `jigsaw_driver.py` is currently calling the "lib" version of `JIGSAW`, which passes everything in memory (using `JIGSAW`'s `API` behind the scenes). The `mesh` object (that's passed by ref here) should hold the output mesh data.

The `cmd-line` version is also available in `python` as `jigsawpy.cmd.jigsaw()`. This passes everything via files (consistent with what's done in `MATLAB`).

`Python` seems to have much better support for external code / libraries than `MATLAB` does, which is why I've been able to support both the `API` and `cmd-line` versions. In `MATLAB` your're stuck with files only!

**Comment by sbrus89** [comment]
@dengwirda, It looks like that worked! Thanks for the help.

**Comment by xylar** [comment]
@sbrus89, I can give you permission to push to my branch, but my preference is for you to just push to your own fork and then make a pull request with yours.  I just wanted to get things going.

Regarding using `jigsawpy.lib.jigsaw()`, I'm guessing a few more explicit commands to write out the meshes would also do the trick.  But given that `cmd` does that anyway, that seems like a fine way to go.

**Comment by xylar** [comment]
@sbrus89, you may be able to remove the explicit calls to save the meshes in `jigsaw_driver.py` if `jigsawpy.cmd.jigsaw()` does this anyway.  Up to you.

**Comment by dengwirda** [comment]
@xylar, @sbrus89: just confirming the workflow for `jigsaw_driver.py` should be:
````
call savemsh() to create geom file
call savemsh() to create hfun file

cmd.jigsaw(), which calls JIGSAW and creates mesh file

mesh file now available for post-processing
````
With the changes yesterday, I think @sbrus89 has this in place.

**Comment by sbrus89** [comment]
@xylar, sounds good I'll go ahead and push to my fork and make the PR from there.

**Comment by sbrus89** [comment]
@xylar and @mark-petersen, should we close this now that #301 is Merged?

**Comment by xylar** [comment]
Yes, thanks.
